traversal/walk.go

```diff
@@ -142,192 +157,267 @@ func (prog Progress) WalkLocal(n datamodel.Node, fn VisitFn) error {
 // An AdvVisitFn is used instead of a VisitFn, so that the reason can be provided.
 func (prog Progress) WalkAdv(n datamodel.Node, s selector.Selector, fn AdvVisitFn) error {
 	prog.init()
-	return prog.walkAdv(n, s, fn)
+	return prog.walkBlock(n, s, fn)
+}
+
+// walkBlock anchors a walk at the beginning of the traversal and at the
+// beginning of each new link traversed. This allows us to do a preload phase if
+// we have a preloader configured.
+func (prog Progress) walkBlock(n datamodel.Node, s selector.Selector, visitFn AdvVisitFn) error {
+	ph := phaseTraverse
+	var budget *Budget
+
+	if prog.Cfg.Preloader != nil {
+		ph = phasePreload
+		budget = prog.Budget.Clone()
+	}
+
+	// First pass.
+	err := prog.walkAdv(ph, n, s, visitFn)
+	if err != nil && (ph != phasePreload || !errors.Is(&ErrBudgetExceeded{}, err)) {
```

Comment:
am I correct reading this that you are making sure the preloader doesn't exceed the load budget, but then you're also going to the actual walk phase even if you hit the link budget? (I assume on the theory that you want to make sure that every link gets explored UNTIL the preload budget is hit?

Also there is a bit of fudging here no? Cause if my linkbudget is 2, and I have three links in root block, so I will load the first two then hit the budget error. However, when I traverse, I will load and traverse the first link. If it has it's own link, I will traverse that, such that when I return back to the root block, by budget is spent, and now, I will never load the 2nd link I preloaded earlier. Is this ok? Expected behavior?

Reply:
So at that point I've loaded 3 links in preloading total (2 on root block and 1 on first sub block), when my budget is 2. This isn't terrible, but it should be spelled out comments on preloading.

Reply:
Yeah, you're right, this is kind of like a best-guess fudge, I'm not sure what else to do about them. An alternative I had for a while was to ignore the budget entirely for the preload pass, but it seems better to at least try.
Will comment, that's a good call, this is far too subtle that it'll likely cause surprise.